### PR TITLE
:white_check_mark: update test

### PR DESCRIPTION
**Pull request by zxs-1024**

### CODE REVIEW

File: test.ts [hunk 1]
```diff
@@ -1,16 +1,49 @@
...
+          type: 'string'
+        }
+      }
+    )
+    expect(options).toEqual({
+      yesterday: true,
+      quiet: 'three'
+    });
+    expect(args).toEqual([]);
+    expect(unknownOptions).toEqual([]);
```
Comment (victorteokw) [inline test.ts:25]: This is a very nice implementation! 👍

File: test.ts [hunk 2]
```diff
@@ -1,16 +1,49 @@
...
   });
 
   it('puts arg tokens into second return value', () => {
-
+    const [options, args, unknownOptions] = parse(
+      ['one', 'two'],
+      {}
```
Comment (victorteokw) [inline test.ts:31]: Let's add some unrelated rule definitions here 🙂

File: test.ts [hunk 3]
```diff
@@ -1,16 +1,49 @@
...
-
+  // TODO: ~unknown 如果传入，则会推入 args，不会推入 unknownOptions
+  it('puts unknown options into third return value', () => {
+    // const [options, args, unknownOptions] = parse(
+    //   ['~unknown'],
+    //   {}
+    // )
+    // expect(options).toEqual({});
+    // expect(args).toEqual([]);
+    // expect(unknownOptions).toEqual(['~unknown']);
```
Comment (victorteokw) [inline test.ts]: For unknown ones, let's define some rules and passing some flags. Only flag that has initial '-' / '--' are considered as options. So all unknown options have a format like '--i-am-unknown', or '-u'.

File: test.ts [hunk 4]
```diff
@@ -39,22 +72,61 @@ describe('does parsing', () => {
 });
 
 describe('supported value types', () => {
+  // TODO: ？？？
   it('boolean without following argument', () => {
```
Comment (victorteokw) [inline test.ts:74]: following argument means "--follow value"
right hand side argument means "--follow=value"

File: test.ts [hunk 5]
```diff
@@ -39,22 +72,61 @@ describe('does parsing', () => {
...
+      quiet: 9
+    });
   });
-
+  // TODO: error "noQuiet": undefined
+  // if use "noquiet" is ok
   it("boolean with prefixed 'no'", () => {
-
+    const [options] = parse(
+      ['--noYesterday', '-noQuiet', 'true'],
```
Comment (victorteokw) [inline test.ts]: This test will not pass, because I haven't implement it. 😄
Prefix no means "--no-yesterday"

File: test.ts [hunk 6]
```diff
@@ -39,22 +72,61 @@ describe('does parsing', () => {
...
+        }
+      }
+    );
+    expect(options).toEqual({
+      'noYesterday': true,
+      'noQuiet': true
+    });
   });
 
+  // TODO: 哈 ?
```
Comment (victorteokw) [inline test.ts]: --quiet true

File: test.ts [hunk 7]
```diff
@@ -39,22 +72,61 @@ describe('does parsing', () => {
...
   it('boolean with rhs true', () => {
 
   });
-
+  // TODO: 哈 ?
```
Comment (victorteokw) [inline test.ts]: --silent false

File: test.ts [hunk 8]
```diff
@@ -39,22 +72,61 @@ describe('does parsing', () => {
...
   it('boolean with rhs false', () => {
 
   });
-
+  // TODO: 哈 ?
   it("boolean with any value that not equal to 'false'", () => {
```
Comment (victorteokw) [inline test.ts]: --help please

File: test.ts [hunk 9]
```diff
@@ -79,24 +151,98 @@ describe('supported value types', () => {
 
   });
 
+  // TODO: error "quiet": NaN "yesterday": NaN
   it('number with following hexadecimal value', () => {
```
Comment (victorteokw) [inline test.ts:294]: hexadecimal value means "0x123abc"

File: test.ts [hunk 10]
```diff
@@ -79,24 +151,98 @@ describe('supported value types', () => {
...
+      }
+    );
+    // expect(options).toEqual({
+    //   yesterday: 255,
+    //   quiet: 255
+    // });
   });
 
   it('number with following decimal value', () => {
-
```
Comment (victorteokw) [inline test.ts:87]: decimal simply means "number system and arithmetic based on the number **ten**"

File: test.ts [hunk 11]
```diff
@@ -79,24 +151,98 @@ describe('supported value types', () => {
...
+    );
+    expect(options).toEqual({
+      yesterday: 9.9,
+      quiet: 9.9
+    });
   });
 
+  // TODO: 设置进制？
   it('number with following octal value', () => {
-
```
Comment (victorteokw) [inline test.ts:91]: octal value is some value like '045'
octal is "number system and arithmetic based on the number **eight**"

File: test.ts [hunk 12]
```diff
@@ -118,58 +264,291 @@ describe('supported value types', () => {
...
+          type: 'string'
+        }
+      }
+    );
+    expect(options).toEqual({
+      quiet: 'three'
+    });
   });
 
+  // TODO: multiple alias in single dash ?
```
Comment (victorteokw) [inline test.ts]: "-hvs" means ["--help", "--version", "--silent"] if you define these rules with the alias names

File: test.ts [hunk 13]
```diff
@@ -118,58 +264,291 @@ describe('supported value types', () => {
...
+    //   yesterday: true,
+    //   quiet: 'three'
+    // });
   });
 });
 
 describe('copies default values for undefined options', () => {
   it('sets true for boolean field', () => {
-
+    const [options] = parse(
```
Comment (victorteokw) [inline test.ts:557]: you need to define a default value in the rules

Comment (victorteokw) [reply test.ts:557]: and do not pass values in argv

File: test.ts [hunk 14]
```diff
@@ -118,58 +264,291 @@ describe('supported value types', () => {
...
+    );
+    expect(options).toEqual({
+      yesterday: true,
+      quiet: true
+    });
   });
 
+  // TODO: error: "quiet": true
   it('sets false for boolean field', () => {
-
```
Comment (victorteokw) [inline test.ts:139]: these and the following are all for default values

File: test.ts [hunk 15]
```diff
@@ -118,58 +264,291 @@ describe('supported value types', () => {
...
+      }
+    );
+    // expect(options).toEqual({
+    //   yesterday: false,
+    //   quiet: false
+    // });
+    // expect(args).toEqual([]);
   });
 
+  // TODO: precedence over？
```
Comment (victorteokw) [inline test.ts:764]: for these, you need to pass at least 3 arguments, see function definition

**Review by victorteokw — CHANGES_REQUESTED** [review]

File: test.ts [hunk 16]
```diff
@@ -40,136 +72,718 @@ describe('does parsing', () => {
 
 describe('supported value types', () => {
   it('boolean without following argument', () => {
-
+    const [options] = parse(
+      ['-stars', '-q'],
```
Comment (victorteokw) [inline test.ts]: Please change '-stars' into '--stars'

File: test.ts [hunk 17]
```diff
@@ -40,136 +72,718 @@ describe('does parsing', () => {
...
+      quiet: true,
+    });
   });
 
   it("boolean with prefixed 'no'", () => {
-
+    const [options] = parse(
+      ['--no-stars', '--no-quiet'],
+      {
+        'no-stars': {
```
Comment (victorteokw) [inline test.ts:100]: --no-stars means value of 'stars' is false,
this test will never pass, please change the function call from `it` to `it.skip`.

File: test.ts [hunk 18]
```diff
@@ -40,136 +72,718 @@ describe('does parsing', () => {
...
+      'stars': false,
+      'moon': false
+    });
   });
 
+  // TODO: it's ok ?
   it("boolean with any value that not equal to 'false'", () => {
-
+    const [options] = parse(
+      ['--stars=true', '--moon=true'],
```
Comment (victorteokw) [inline test.ts]: let's use any value other than true and false to test. --stars=whatever --moon=however

File: test.ts [hunk 19]
```diff
@@ -40,136 +72,718 @@ describe('does parsing', () => {
...
+    expect(options).toEqual({
+      'stars': true,
+      'moon': true
+    });
   });
 
   it('number with rhs hexadecimal value', () => {
-
+    const [options] = parse(
+      ['--starfish=1314', '--hippocampus=1024'],
```
Comment (victorteokw) [inline test.ts]: --starfish=0x522 --hippocampus=0x400

File: test.ts [hunk 20]
```diff
@@ -40,136 +72,718 @@ describe('does parsing', () => {
...
+        },
+        'hippocampus': {
+          alias: 'q',
+          desc: 'the quiet one.',
+          type: 'number'
+        }
+      }
+    );
+    expect(options).toEqual({
+      'starfish': 0o2442,
```
Comment (victorteokw) [inline test.ts]: decimal is ten based. I don't what you to do math arithmetic operations

File: test.ts [hunk 21]
```diff
@@ -40,136 +72,718 @@ describe('does parsing', () => {
...
+      'hippocampus': 0o2000
+    });
   });
 
   it('number with rhs octal value', () => {
-
+    const [options] = parse(
+      ['--starfish=1314', '--hippocampus=1024'],
```
Comment (victorteokw) [inline test.ts]: octal is eight based. octal numbers starts with a prefix 0

File: test.ts [hunk 22]
```diff
@@ -40,136 +72,718 @@ describe('does parsing', () => {
...
+    expect(options).toEqual({
+      'starfish': 1314,
+      'hippocampus': 1024
+    });
   });
 
   it('number with rhs binary value', () => {
-
+    const [options] = parse(
+      ['--starfish=1314', '--hippocampus=1024'],
```
Comment (victorteokw) [inline test.ts]: binary value is like 0b1001010

**Review by victorteokw — CHANGES_REQUESTED** [review]
Hi @zhanghao-zhoushan, 
Very great job! And it's really a lot of works. 👍👍
Please update according to the comments. 
The last three cases, I'll just modify.
And after this, can you squash these commits into one and name it "implement tests"?

Thank you

File: test.ts (hunk 20 above)
Comment (zxs-1024) [reply test.ts]: It's a mistake.

**Review by victorteokw — APPROVED** [review]

**Pull request merged**